### PR TITLE
POPM Implementation

### LINKED ISSUE #44: POPM
Popularimeter

http://id3.org/id3v2.3.0#sec4.18

### PR DESCRIPTION
**Pull request by dhulihan**
Resolves #44 . Supports Popularimeter Rating, Email, and Play Counter.

https://id3.org/id3v2.3.0#Popularimeter

### CODE REVIEW

File: popularimeter_frame.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+const (
+	// maxCounterBytes is the maximum number of bytes that a counter can be
+	// when parsed or written. Counter size can currently not be larger than go
+	// uint64 (8 bytes).
+	maxCounterBytes = 8
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:13]: good idea, I'll try `big.Int`

File: popularimeter_frame.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+	// The rating is 1-255 where 1 is worst and 255 is best. 0 is unknown.
+	Rating uint8
+
+	// Counter is the number of times this file has been played by this email.
+	Counter uint64
+}
+
+// AddPopularimeterFrame adds a popularimeter frame to tag.
+func (tag *Tag) AddPopularimeterFrame(pf PopularimeterFrame) {
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:29]: that works, I'll remove

File: popularimeter_frame.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	pf, ok := f.(PopularimeterFrame)
+
+	if !ok {
+		return PopularimeterFrame{}
+	}
+
+	return pf
+}
+
+// UniqueIdentifier uniquely identifies a PopularimeterFrame.
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:50]: np, I usually add documentation to public functions to satisfy golint conventions

File: popularimeter_frame.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+// Size returns the size of the frame body.
+func (pf PopularimeterFrame) Size() int {
+	ratingSize := 1
+	return (len(pf.Email) + len(EncodingISO.TerminationBytes) + ratingSize + len(pf.counterBytes()))
+}
+
+// get a byte slice that represents the counter
+func (pf PopularimeterFrame) counterBytes() []byte {
+	b := make([]byte, maxCounterBytes)
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:63]: thanks, good suggestion

File: popularimeter_frame.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+	n := binary.PutUvarint(b, pf.Counter)
+	if n <= 0 {
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:66]: good catch, `binary.PutUvarint` will return 0. mixed this up with `binary.Uvarint`, which will return negative numbers :)

File: popularimeter_frame.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		bw.WriteByte(pf.Rating)
+		bw.Write(pf.counterBytes())
+	})
+}
+
+func parsePopularimeterFrame(br *bufReader) (Framer, error) {
+	email := br.ReadText(EncodingISO)
+	rating := br.ReadByte()
+
+	counterBytes := br.Next(maxCounterBytes)
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:86]: you're right, we should fully support the spec here, not just a max of 8 bytes

File: popularimeter_frame.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+func (pf PopularimeterFrame) Size() int {
+	ratingSize := 1
+	return (len(pf.Email) + len(EncodingISO.TerminationBytes) + ratingSize + len(pf.counterBytes()))
+}
+
+// get a byte slice that represents the counter
+func (pf PopularimeterFrame) counterBytes() []byte {
+	b := make([]byte, maxCounterBytes)
+
+	n := binary.PutUvarint(b, pf.Counter)
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:63]: Ah this poses an interesting challenge. If the number of bytes for play counter is variable (at least 4 bytes or possibly more), we'll need to know how big the frame body is at the time of parsing, so we can properly compute `Size()`. It looks like frame body size is only accessible here:

https://github.com/bogem/id3v2/blob/master/parse.go#L85

How do you think we should go about figuring out how many counter bytes there are when we parse? We could peek ahead in the `bufReader`, hmm.

[the remark this answers is not in the document]
Comment (bogem) [reply popularimeter_frame.go:63]: You don't have to know the frame size.

There is a same challenge in [picture frame](https://github.com/bogem/id3v2/blob/master/picture_frame.go#L50). You just use `bufReader.ReadAll` and it reads till the end of frame. This bufReader wraps the [limited reader](https://github.com/bogem/id3v2/blob/master/parse.go#L90) that limits to the body size of frame.

File: popularimeter_frame.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go]: Updated to use `big.Int` and `br.ReadAll()` to read variable-length counter. Thanks for the suggestion!

File: tag.go [hunk 8]
```diff
@@ -257,6 +257,17 @@ func (tag *Tag) SetGenre(genre string) {
...
+// Popularimeter returns an email address string, and a rating represented by uint8. 1 is bad, 255 is good, and 0 is unknown.
+func (tag *Tag) Popularimeter() (string, uint8, uint64) {
+	p := tag.GetPopularimeterFrame()
+	return p.Email, p.Rating, p.Counter
+}
+
+// SetPopularimeter sets an email and rating.
+func (tag *Tag) SetPopularimeter(email string, rating uint8, counter uint64) {
+	tag.AddPopularimeterFrame(PopularimeterFrame{Email: email, Rating: rating, Counter: counter})
+}
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply tag.go]: removed

File: popularimeter_frame.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go]: alrighty, updated to use `big.Int` to handle variable length counter bytes.

File: popularimeter_frame_test.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	}
+
+	if actual.Rating != expected.Rating {
+		t.Errorf("Expected rating: %v, got: %v", expected.Rating, actual.Rating)
+	}
+
+	if actual.Counter != expected.Counter {
+		t.Errorf("Expected counter: %d, got: %d", expected.Counter, actual.Counter)
+	}
+}
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame_test.go]: moved test to `parse_test.go`.

File: popularimeter_frame.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go]: thanks for the tip, updated to use `bufReader.ReadAll`. That's a handy function :) .

**Comment by dhulihan** [comment]
Alrighty, @bogem , thanks for the feedback. I made some changes based on your commentary, let me know if I missed anything.

File: tag_test.go [hunk 10]
```diff
@@ -146,7 +155,7 @@ func TestCountLenSize(t *testing.T) {
 	}
 
 	// Check len of tag.AllFrames().
-	if len(tag.AllFrames()) != 11 {
+	if len(tag.AllFrames()) != 12 {
 		t.Errorf("Expected: %v, got: %v", 11, len(tag.AllFrames()))
 	}
```
Comment (bogem) [inline tag_test.go:161]: I think we should also add a test that makes sure that counter is at least 4 bytes. You can make a new file (e.g. `popularimeter_frame_test.go`) and place this test:
```go
package id3v2

import (
	"bytes"
	"math/big"
	"testing"
)

// Make sure that counter of popularimeter is at least 4 bytes even if it's small number.
func TestPopularimeterFrameSmallCounter(t *testing.T) {
	popmFrame := PopularimeterFrame{
		Email:   "foo@bar.com",
		Rating:  1,
		Counter: big.NewInt(1),
	}

	expectedBodyLength := len(popFrame.Email) + 1 + 1 + 4

	buf := new(bytes.Buffer)
	written, err := popmFrame.WriteTo(buf)
	if err != nil {
		t.Fatalf("Error by writing: %v", err)
	}

	if written != int64(expectedBodyLength) {
		t.Fatalf("Expected popularimeter frame body length: %v, got: %v", expectedBodyLength, written)
	}

	expectedCounter := []byte{0, 0, 0, 1}
	gotCounter := buf.Bytes()[expectedBodyLength-4:]
	if !bytes.Equal(expectedCounter, gotCounter) {
		t.Fatalf("Expected popularimeter counter: %v, got: %v", expectedCounter, gotCounter)
	}
}
```

File: popularimeter_frame.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
+package id3v2
+
+import (
+	"io"
+	"math/big"
+)
+
+// PopularimeterFrame structure is used for Popularimeter (POPM).
+// https://id3.org/id3v2.3.0#Popularimeter
+type PopularimeterFrame struct {
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:10]: good thinking, I added a get example as well.

File: popularimeter_frame.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (pf PopularimeterFrame) Size() int {
+	ratingSize := 1
+	return len(pf.Email) + 1 + ratingSize + len(pf.counterBytes())
+}
+
+// counterBytes returns a byte slice that represents the counter.
+func (pf PopularimeterFrame) counterBytes() []byte {
+	counterBytes := pf.Counter.Bytes()
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go:32]: updated

File: tag_test.go (hunk 10 above)
Comment (dhulihan) [reply tag_test.go:161]: good thinking, added. Thanks for fleshing out this test.

File: parse_test.go [hunk 13]
```diff
@@ -186,6 +188,32 @@ func compareTXXXFrames(actual, expected UserDefinedTextFrame) error {
 	return nil
 }
 
+func testPopularimeterFrame(t *testing.T, tag *Tag) {
+	expected := PopularimeterFrame{
+		Email:   "foo@bar.com",
+		Rating:  uint8(128),
+		Counter: big.NewInt(10000000000000000),
+	}
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply parse_test.go]: updated

File: popularimeter_frame.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+	// spec requires at least 4 bytes for play counter, pad if necessary
+	bytesNeeded := 4 - len(pf.Counter.Bytes())
```
[the remark this answers is not in the document]
Comment (dhulihan) [reply popularimeter_frame.go]: thanks, good catch

**Comment by dhulihan** [comment]
Thanks for the great feedback and working with me on this @bogem !

**Pull request merged**